### PR TITLE
FIX-#3376: fix handling of groupby-dict aggregation of 'by' cols

### LINKED ISSUE #3376: Groupby dictionary aggregation works incorrect in cases when 'by' and columns to aggregate has intersection
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:

Ubuntu 20.04.3 LTS

- **Modin version** (`modin.__version__`):

0.10.2+2.g253ecd98.dirty

- **Python version**:

Python 3.8.6

- **Code we can use to reproduce**:

```python
import modin.pandas as pd

df = pd.DataFrame({
    "col0": [ 1,  2,  1,  2],
    "col1": [11, 12, 13, 14],
    "col2": [ 4,  4,  4,  4],
    "col3": [ 1,  1,  1,  1]
})

print(df.shape)
gb = df.groupby("col0")
result = gb.agg({"col0": "count"})
string = "%s" % repr(result)
print(string)
```

### Describe the problem

Code above works on Pandas but produces a `ValueError: No objects to concatenate` exception on Modin on Python and Ray engines (those that I tried). I traced the problem down to these lines
```python
        if drop and len(df.columns.intersection(by_part)) > 0:
            df.drop(columns=by_part, errors="ignore", inplace=True)
```
that are located inside of `reduce` function. As a result of these lines `df` that contains only one column `col0` gets empty because this column is dropped. I am not sure whether it is intentional, it doesn't look right for me. @dchigarev please comment.

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

Previously, there was no ability to select  the same columns that the frame was grouped on in a dict aggregation because the following code was dropping all of the intersections of 'by' and the aggregated columns in order to avoid naming conflicts:
https://github.com/modin-project/modin/blob/ddd4afe487d1aee19426cd454e4412a17f1a555c/modin/core/dataframe/algebra/groupby_function.py#L196-L197
This PR changes this logic, by not dropping the columns that were explicitly selected for aggregation and also introducing a common logic of handling the `as_index=False` parameter and all of the derived naming conflicts for any groupby implementation (`GroupBy.handle_as_index` method).

Please visit the original issue (#3376) for a more precise problem description.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3376 
- [ ] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [inline modin/core/dataframe/algebra/default2pandas/groupby.py:212]: avoiding errors when `by` is a Series:
```python
>>> pandas.Series([1, 2, 3]).squeeze(axis=1)
ValueError: No axis named 1 for object type Series
```

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [inline modin/core/dataframe/algebra/default2pandas/groupby.py:239]: this logic is now performed by `handle_as_index`

**Comment by dchigarev** [comment]
@modin-project/modin-core a reminder to review

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (vnlitvinov) [inline modin/core/dataframe/algebra/default2pandas/groupby.py:406]:
```suggestion
            internal_by_cols[~internal_by_cols.str.startswith("__reduced__", na=False)]
```
or something like that. Do not use regexps unless really necessary, in this case for performance reasons (however, some measurements would be good to prove me right or wrong tho)

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (vnlitvinov) [inline modin/core/dataframe/algebra/default2pandas/groupby.py:416]:
```suggestion
        if selection is not None and not isinstance(selection, pandas.Index):
            selection = pandas.Index(selection)
```

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (vnlitvinov) [inline modin/core/dataframe/algebra/default2pandas/groupby.py:434]: I suggest to rewrite this in a normal `if..else` way, current one looks overcomplicated.

Don't try to fit everything in a single statement, it makes things harder to understand in a lot of cases.

```python
if selection is None:
    if internal_by_cols.nlevels != result_cols.nlevels:
        cols_to_insert = internal_by_cols.copy()
    else:
        cols_to_insert = frozenset(internal_by_cols) - frozenset(result_cols)
else:
    cols_to_insert = frozenset(internal_by_cols) - frozenset(selection)
```
or something along this lines

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (vnlitvinov) [inline modin/core/dataframe/algebra/default2pandas/groupby.py:442]:
```suggestion
                else (frozenset(cols_to_insert) & frozenset(result_cols))
```

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (vnlitvinov) [inline modin/core/dataframe/algebra/default2pandas/groupby.py:453]: again, I suggest rewriting this to a simpler outer if-else form

File: modin/core/dataframe/algebra/groupby.py
Comment (vnlitvinov) [reply modin/core/dataframe/algebra/groupby.py:124]: please add this info as a comment, something like
```python
# cls.try_filter_dict() should go before insertion of missing 'by' columns, otherwise
# if 'by' is presented in the agg dict, the aggregation of 'by' cols would be computed in every partition
```

File: modin/pandas/test/test_groupby.py
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1809]: very nice, excellent to see test explanations!

File: modin/pandas/test/test_groupby.py
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1808]:
```suggestion
        3. Compute GroupBy result with the ``as_index=False`` parameter via pandas as the reference result.
```

File: modin/pandas/test/test_groupby.py
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1813]: shouldn't this be an assertion that the function isn't actually testing anything (called uselessly)?

File: modin/pandas/test/test_groupby.py
Comment (vnlitvinov) [reply modin/pandas/test/test_groupby.py:1823]: If you feel that you need to explain code changes to fellow reviewers, please _avoid_ doing so by PR comments.
Nobody would see those after a PR is merged.

If you feel an urge to explain - do it via in-code comments!

File: modin/pandas/test/test_groupby.py
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1854]: please don't save a few typed letters and spell the words in the full - I suggest replacing `res` with `result` and `ref` with `reference` or smth similar (here and in other places of this function)

File: modin/pandas/test/test_groupby.py
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1843]: do we have a test for `.validate_by()`?

File: modin/pandas/test/test_groupby.py
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1873]: I saw this pattern at least 4 or 5 times already... I feel that it should be a helper function or something.

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [reply modin/core/dataframe/algebra/default2pandas/groupby.py:406]: `startswith` is slightly faster indeed:
```
startswith:
        Total time of 10 runs: 6.242122787982225; Mean time: 0.6242122787982225
regexp:
        Total time of 10 runs: 7.681726414011791; Mean time: 0.7681726414011791
```

<details><summary>reproduced</summary>

```python
import pandas
import timeit
import random

random.seed(42)
n_elements = 1_000_000
n_runs = 10

reduced_columns_percentage = 0.1
almost_reduced_columns_percentage = 0.2
regular_columns_percentage = 1 - (
    reduced_columns_percentage + almost_reduced_columns_percentage
)

reduced = [
    f"__reduced__someth{i}ngsomething"
    for i in range(int(n_elements * reduced_columns_percentage))
]
almost_reduced = [
    f"__reduced_someth{i}ngsomething"
    for i in range(int(n_elements * almost_reduced_columns_percentage))
]
regular = [
    f"someth{i}ngsomething" for i in range(int(n_elements * regular_columns_percentage))
]

index = reduced + almost_reduced + regular
random.shuffle(index)

index = pandas.Index(index)

startswith_time = timeit.timeit(
    lambda: index.str.startswith("__reduced__", na=False), number=n_runs
)
regexp_time = timeit.timeit(
    lambda: index.str.match(r"__reduced__.*", na=False), number=n_runs
)

print(
    f"startswith:\n\tTotal time of {n_runs} runs: {startswith_time}; Mean time: {startswith_time / n_runs}"
)
print(f"regexp:\n\tTotal time of {n_runs} runs: {regexp_time}; Mean time: {regexp_time / n_runs}")
```

</details>

P.S. Timings of half a second for this filtering may scare, but note, that in the repro I filter a million values when in the actual code the filtered object is an Index, containing column names to group on, its length is unlikely to be big.

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [reply modin/core/dataframe/algebra/default2pandas/groupby.py]: Created an issue and referenced it in the in-code comment (#3739)

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [reply modin/core/dataframe/algebra/default2pandas/groupby.py:211]: No, we shouldn't. This is a kernel function, that only operates pandas objects, this isn't mentioned in the function's doc-string though.

As a proof, I can refer to the `build_groupby` fn, whose [docstring says ](https://github.com/dchigarev/modin/blob/f203dff378bb48d8f378ce9d48d79242841a8ba1/modin/core/dataframe/algebra/default2pandas/groupby.py#L283-L286) that it returns a function taking only pandas DataFrames.

File: modin/core/dataframe/algebra/groupby.py
Comment (dchigarev) [reply modin/core/dataframe/algebra/groupby.py:127]: Done, added the following in-code comment:
```python
        # We have to filter func-dict BEFORE inserting broadcasted 'by' columns
        # to avoid multiple aggregation results for 'by' cols in case they're
        # present in the func-dict:
```

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [reply modin/core/dataframe/algebra/default2pandas/groupby.py]: Our `pydocstyle` check claims that the first word in the docstring must be a verb. Changed to the:
"Compute hints to process ``as_index=False`` parameter for the GroupBy result."

File: modin/pandas/test/test_groupby.py
Comment (dchigarev) [reply modin/pandas/test/test_groupby.py]: you mean `pytest.skip` or something similar?

added this for now:
```python
    if by_length == 0:
        pytest.skip("No keys to group on was passed, skipping the test.")
```

File: modin/pandas/test/test_groupby.py
Comment (dchigarev) [reply modin/pandas/test/test_groupby.py]: wrapped it into the `handle_as_index_for_dataframe` function

File: modin/pandas/test/test_groupby.py
Comment (dchigarev) [reply modin/pandas/test/test_groupby.py]: no, we don't. My feeling is that it should be done separately from these changes, but added the test in this PR just in case.

File: modin/pandas/test/test_groupby.py
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py]: why did you remove usage of named columns here?

File: modin/pandas/test/test_groupby.py
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1836]:
```suggestion
    by_length = internal_by_length + external_by_length
```
isn't it the same, huh?

File: modin/pandas/test/test_groupby.py
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1838]:
```suggestion
        pytest.skip("No keys to group on were passed, skipping the test.")
```

File: modin/pandas/test/test_groupby.py
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1870]: are you telling me this clever way is better than
```suggestion
    internal_by = df.columns[range(-internal_by_length // 2, internal_by_length // 2)].tolist()
```
? sure, my suggestion takes consecutive columns from the beginning and end and in different order, but does it matter?

File: modin/pandas/test/test_groupby.py
Comment (vnlitvinov) [inline modin/pandas/test/test_groupby.py:1919]: what you're checking and what you're reporting are two different things, pls adjust the message

File: modin/core/storage_formats/pandas/query_compiler.py
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py:2790]: This won't work for numpy-like arrays (e.g. pandas Index) since its `__nonzero__` raises an exception:

<details><summary>Example</summary>

```python
>>> if np.array([1, 2]):
...     print("some")
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
>>> if np.array([]):
...     print("some")
...
__main__:1: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
>>> if pandas.Index([1, 2]):
...     print("some")
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "lib\site-packages\pandas\core\indexes\base.py", line 2810, in __nonzero__
    f"The truth value of a {type(self).__name__} is ambiguous. "
ValueError: The truth value of a Int64Index is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
>>> if pandas.Index([]):
...     print("some")
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "lib\site-packages\pandas\core\indexes\base.py", line 2810, in __nonzero__
    f"The truth value of a {type(self).__name__} is ambiguous. "
ValueError: The truth value of a Index is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```

</details>

File: modin/core/storage_formats/pandas/query_compiler.py
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py:2717]: answered [here](https://github.com/modin-project/modin/pull/3592#discussion_r758254191)

File: modin/pandas/test/test_groupby.py
Comment (dchigarev) [reply modin/pandas/test/test_groupby.py]: I did this because we don't know the exact column names of the testing frame here, so I just used their numerical indices to compute the actual labels when the info about the frame will be available.

File: modin/core/dataframe/algebra/default2pandas/groupby.py
Comment (dchigarev) [inline modin/core/dataframe/algebra/default2pandas/groupby.py]: There is no way of getting a different amount of levels for the 'by' columns and for the resulting ones unless it's a renaming aggregation (handled separately in the `selection is not None` branch).

This if-branch was originally misported from the [`qc.groupby_agg`](https://github.com/modin-project/modin/blob/045eb9353f234f67e5bf35272ea98aac9c05cdf2/modin/core/storage_formats/pandas/query_compiler.py#L2764-L2775) where there was no distinguishing between `selection` and `non-selection` aggregations.

File: modin/pandas/test/test_groupby.py
Comment (dchigarev) [inline modin/pandas/test/test_groupby.py:1857]: since picking of the 'by' columns now starts from negative indices (and not always including '0' or positive ones), I've changed the index of the categorical column as well, so there's always be a categorical col in 'by'.

File: modin/pandas/test/test_groupby.py
Comment (dchigarev) [reply modin/pandas/test/test_groupby.py]: changed the check to the `assert type(obj1) == type(obj2)`

**Review by vnlitvinov — DISMISSED** [review]
LGTM assuming CI passes.

**Review by vnlitvinov — APPROVED** [review]
_Re-approving after merge commit._

LGTM assuming CI passes.

**Comment by dchigarev** [comment]
@vnlitvinov CI is green

File: modin/pandas/test/test_groupby.py
Comment (YarShev) [inline modin/pandas/test/test_groupby.py:1814]: @modin-project/modin-core, @modin-project/modin-omnisci, let's not use `backend` as a standalone word. It is a vague term. We have more clear words now: storage format, engine and execution.

File: modin/pandas/test/test_groupby.py
Comment (YarShev) [inline modin/pandas/test/test_groupby.py:1882]: I would prefer to change all entries having `*backend*`.

Comment (dchigarev) [reply modin/pandas/test/test_groupby.py:1882]: Created an issue for this (#3787). Will fix soon

Comment (YarShev) [reply modin/pandas/test/test_groupby.py:1882]: Thanks!

**Pull request merged**
